<a href="https://colab.research.google.com/github/Alecs400/SQL./blob/main/Sprint_8_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [72]:
import pandas as pd

# Carga del archivo Excel de empleados
df = pd.read_excel(
    "/content/drive/MyDrive/Pandas_Sprint_8/sprint8_complex.xlsx"
)

# Exploración inicial (cambiado a 5 o 10 para no saturar la pantalla con 1000 filas)
display(df.head(10))

,Dades de l'enquesta de treballadors 2025,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19
0,Generades aleatòriament per seguretat,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
3,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
4,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
5,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
6,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
7,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
8,5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1.0,8.0,1978.0,D,"2,069 €",NaN,True,Grup B,Cap de projecte,0.0,NaN,NaN,NaN,19.9
9,6,David,Gómez Vila,36998038E,Espanya,Palma,25.0,9.0,1971.0,D,"2,349 €",NaN,True,Grup B,Cap de projecte,NaN,True,6196.0,10.6,17.8


In [73]:
import pandas as pd

# Lee el archivo directamente desde la carpeta temporal de Colab
df = pd.read_excel("sprint8_complex.xlsx", engine="openpyxl")

display(df.head(10))

,Dades de l'enquesta de treballadors 2025,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19
0,Generades aleatòriament per seguretat,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
3,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
4,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
5,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
6,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
7,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
8,5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1.0,8.0,1978.0,D,"2,069 €",NaN,True,Grup B,Cap de projecte,0.0,NaN,NaN,NaN,19.9
9,6,David,Gómez Vila,36998038E,Espanya,Palma,25.0,9.0,1971.0,D,"2,349 €",NaN,True,Grup B,Cap de projecte,NaN,True,6196.0,10.6,17.8


#Exercici 1: Exploració inicial del dataset


# » Quantes files i columnes té el dataset?


In [74]:
df.shape

(1010, 20)

#Quin tipus de dades conté cada columna?

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Data columns (total 20 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Dades de l'enquesta de treballadors 2025  1008 non-null   object
 1   Unnamed: 1                                1007 non-null   object
 2   Unnamed: 2                                1007 non-null   object
 3   Unnamed: 3                                1007 non-null   object
 4   Unnamed: 4                                1007 non-null   object
 5   Unnamed: 5                                1007 non-null   object
 6   Unnamed: 6                                1007 non-null   object
 7   Unnamed: 7                                1007 non-null   object
 8   Unnamed: 8                                1007 non-null   object
 9   Unnamed: 9                                967 non-null    object
 10  Unnamed: 10                               1007 n

#Hi ha files duplicades?

In [76]:
df.duplicated().sum()

np.int64(0)

El archivo no tiene filas completamente duplicadas. Sin embargo, si se analiza por identificadores personales como el DNI, sí que se encuentran 15 duplicidades de documentos de identidad repetidos.



# Les dades semblen coherents a primera vista?


A simple vista no son totalmente coherentes, ya que el archivo presenta varios problemas evidentes de consistencia en sus datos que saltan a la vista al inspeccionarlo. Por ejemplo:

**Formatos de texto y codificación corruptos:** Hay caracteres mal codificados, lo que indica problemas con la codificación de caracteres.

**Inconsistencias en los tipos de datos:** Los salarios se almacenan como texto mezclando números con símbolos de moneda y formatos mixtos.

**Valores nulos y contradicciones lógicas:** Hay columnas relacionadas como Fills y No Fills o Te_cotxe y Km_anuals que presentan valores vacíos cruzados que requerirán limpieza y transformación para poder trabajar con ellos de forma coherente.

**Duplicidades de identidad:** Existen DNIs repetidos en el dataset.

#» Detectes possibles problemes o inconsistències evidents?


Por añadir más incoherencias / inconsistencias que he encontrado:

**Valores numéricos ilógicos o extremos:** Se aprecian anomalías como valores negativos en variables donde no debería ocurrir (por ejemplo, en los kilómetros anuales recorridos) y fechas de nacimiento con años totalmente erroneos, etc.

**Gran cantidad de valores nulos:** Dependiendo de la columna, hay una dispersión notable de datos faltantes que obligarán a decidir si se eliminan filas, se imputan valores o se tratan de otra forma durante la fase de limpieza.

#Exercici 2: Validacions bàsiques de les dades i decisions d'eliminació


In [77]:
import pandas as pd

# Leemos el archivo localmente con header=2
df = pd.read_excel("sprint8_complex.xlsx", header=2, engine="openpyxl")

# Para comprobar que se ha cargado correctamente:
display(df.head(10))

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19
0,NaN,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
1,0.0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
2,1.0,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
3,2.0,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
4,3.0,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
5,4.0,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
6,5.0,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1.0,8.0,1978.0,D,"2,069 €",NaN,True,Grup B,Cap de projecte,0.0,NaN,NaN,NaN,19.9
7,6.0,David,Gómez Vila,36998038E,Espanya,Palma,25.0,9.0,1971.0,D,"2,349 €",NaN,True,Grup B,Cap de projecte,NaN,True,6196.0,10.6,17.8
8,7.0,Carmen,López,47376585K,Espanya,Barcelona,8.0,9.0,1962.0,H,"2,855 €",True,NaN,Grup C,Responsable de vendes,4.0,True,20000,5.9,15.4
9,540.0,Nora,Andersen Pedersen,37289414J,Noruega,Stavanger,8.0,3.0,1963.0,D,"1,338 €",NaN,True,Grup B,Analista junior,0.0,False,NaN,NaN,7.6


#» Existeixen identificadors únics (DNI) per a cada persona?


**Nota: Primero de todo, cargamos el Dataset y corregimos el problema de los encabezados, para poder filtrar a posteriori por columna cuántos DNI hay repetidos y demás.**

In [78]:
import pandas as pd

# Cargamos el Excel localmente con skiprows=3 y el motor openpyxl
df = pd.read_excel(
    "sprint8_complex.xlsx", skiprows=3, engine="openpyxl"
)
df = df.drop(columns=["Unnamed: 0"])

# Convertir de manera masiva cualquier texto mal codificado
for col in df.select_dtypes(include=["object"]).columns:
  df[col] = (
      df[col]
      .astype(str)
      .apply(
          lambda x: (
              x.encode("latin1", errors="ignore")
              .decode("utf-8", errors="ignore")
              if "Ã" in x or "Ô" in x
              else x
          )
      )
  )

display(df.head(10))

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,nan,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",nan,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,nan,Grup A,Tècnic IT,4.0,nan,NaN,NaN,16.7
3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,nan,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",nan,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1.0,8.0,1978.0,D,"2,069 €",nan,True,Grup B,Cap de projecte,0.0,nan,NaN,NaN,19.9
6,David,Gómez Vila,36998038E,Espanya,Palma,25.0,9.0,1971.0,D,"2,349 €",nan,True,Grup B,Cap de projecte,NaN,True,6196.0,10.6,17.8
7,Carmen,López,47376585K,Espanya,Barcelona,8.0,9.0,1962.0,H,"2,855 €",True,nan,Grup C,Responsable de vendes,4.0,True,20000.0,5.9,15.4
8,Nora,Andersen Pedersen,37289414J,Noruega,Stavanger,8.0,3.0,1963.0,D,"1,338 €",nan,True,Grup B,Analista junior,0.0,False,NaN,NaN,7.6
9,Alba,García,75181648S,Espanya,Las Palmas de Gran Canaria,32.0,0.0,1940.0,H,"1,450 €",True,nan,Grup A,Senior analyst,4.0,False,NaN,NaN,20.1


**Ahora sí, comprobamos cuantos DNI hay repetidos.**

In [79]:
# 1. Contamos cuantas veces sale cada DNI
conteo = df['DNI'].value_counts()

# 2. Filtramos solo los que se repiten (> 1)
duplicats = conteo[conteo > 1]

# 3. Imprime el total y cuales son
print(f"Nombre de DNI repetits: {len(duplicats)}")
print("Els DNI que es repeteixen són:")
print(duplicats)

print("-" * 50) # Una linea separadora visual

# 4. Extraer e imprimir
dni_repetits = duplicats.index
files_duplicades = df[df['DNI'].isin(dni_repetits)].sort_values(by='DNI')

print("-------- Listado de DNIs duplicados --------")
display(files_duplicades)

Nombre de DNI repetits: 9
Els DNI que es repeteixen són:
DNI
90585882A    6
16618423V    4
55283701C    2
55297384F    2
15909601K    2
52590001R    2
97954019T    2
54013571W    2
49529438G    2
Name: count, dtype: int64
--------------------------------------------------
-------- Listado de DNIs duplicados --------


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
604,Laia,Navarro,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,"1,676 €",True,False,Grup A,Analista,1.0,True,50.0,NaN,14.2
250,Laia,Navarro Vila,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,1235 euros,True,True,Grup D,Analista junior,0.0,False,NaN,NaN,10.3
759,Oriol,Gómez,16618423V,Espanya,Sevilla,10.0,1.0,1968.0,nan,"2,000 €",nan,nan,Grup A,Responsable de vendes,5.0,nan,NaN,NaN,20.7
758,Oriol,Gómez,16618423V,Espanya,Màlaga,10.0,1.0,1968.0,H,1879 euros,True,nan,Grup B,Data Analyst,4.0,False,NaN,NaN,NaN
754,Oriol,Gómez,16618423V,Espanya,Barcelona,10.0,1.0,1968.0,Home,2.358 €,True,nan,nan,nan,4.0,nan,NaN,NaN,NaN
654,Oriol,Gómez,16618423V,Espanya,Madrid,10.0,1.0,1968.0,Home,1731 euros,True,nan,Grup B,Data analyst,4.0,nan,NaN,NaN,NaN
692,Alexia,Pérez Díaz,49529438G,Espanya,Màlaga,12.0,10.0,1981.0,Dona,1663€,nan,True,Grup C,Cap de projecte,NaN,True,80000.0,10.7,16.9
558,Alexia,Pérez Díaz,49529438G,Espanya,Màlaga,12.0,10.0,1981.0,Dona,"1,176 €",nan,True,Grup A,Data Analyst,0.0,False,NaN,NaN,38.0
757,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,nan,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,nan,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9


**Conclusión: ¿Cuántos DNI hay repetidos? La respuesta es, 9 DNIs repetidos. En el listado anterior podemos ver todos los perfiles / registros afectados por estos DNI redundantes**

#» Què faràs amb les files duplicades o registres repetits?


He detectado que hay 3 tipos diferentes de datos duplicados. Los casos son:

**Caso A**: Duplicados exactos o copias técnicas (Ej: da Karlsen, Claudia Gil Sara Serra, Alba o Oriol Ferrer, entre otros).

- Qué haría: Eliminar una de las copias y quedarme con la otra.

- Por qué: Tienen exactamente el mismo nombre, apellidos, DNI y datos prácticamente idénticos. Son errores clásicos donde el registro se ha duplicado en la base de datos.

**Caso B**: El caso de Laia Navarro, Alexia López o Oriol Gómez.

- Qué haría: Revisaría si la fila es 100% idéntica. Si lo es, eliminar uno de los dos índice. Si hay pequeñas variaciones, tendría que decantarme por eliminar el registro que parezca más corrupto o un error tipográfico más evidente.

**Caso C**: Mismo DNI pero personas diferentes: Marta Ruiz o Guillem Pujol, entre otros.

- Qué haría: No eliminar las filas de golpe, porque perderíamos info, pero sí limpiaría el DNI (poniéndolo a NaN).

- Por qué: Tienen nombres y apellidos completamente diferentes; por lo tanto, es un error grave de asignación de DNI en la base de datos original. Si elimináramos las filas, nos cargaríamos el sueldo y la ciudad de Marta y Guillem injustamente.

**Caso A: Eliminar duplicado**

In [80]:
df = df.drop_duplicates()

**Casso B: Evaluar diferencias**

In [81]:
df = df.drop(index=250) # He optado por esta solución porque Laia Navarro tiene un duplicado con un rol y sueldo superior a los duplicados.
df = df.drop(index=558) # He optado por esta solución porque Alexia López tiene un duplicado con un rol y sueldo superior a los duplicados.
df = df.drop(index=[25,913]) # He optado por esta solución porque Sara Serra tiene un duplicado sueldo superior a los duplicados.
df = df.drop(index=[758, 754, 654]) # Con Oriol Gómez borro los registros que tiene 4 hijos, en vez de 5.

**Caso C: Gestionar los casos especials con DNI compartido por personas diferentes**

In [82]:
df.loc[142, 'DNI'] = None
df.loc[89, 'DNI'] = None
df.loc[281, 'DNI'] = None
df.loc[114, 'DNI'] = None
df.loc[104, 'DNI'] = None

In [83]:

# Paso 1: Sacar los nulos del DataFrame
df_sense_nuls = df.dropna(subset=['DNI'])

# Paso 2: Mostrar los duplicados totales. Sin None.
duplicats_dni = df_sense_nuls[df_sense_nuls['DNI'].duplicated(keep=False)].sort_values(by='DNI')
display(duplicats_dni)

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat


**DNIs con letra minúscula**

In [84]:
# Cuenta cuantos DNI tienen alguna letra minúscula
compte_minuscules = (
    df["DNI"].astype(str).str.contains(r"[a-z]", regex=True).sum()
)

print(f"Nombre de DNI amb lletres minúscules: {compte_minuscules}")
# Muestra solo las filas con DNI en minúscula
dni_amb_minuscules = df[df["DNI"].astype(str).str.contains(r"[a-z]", regex=True)]
print(dni_amb_minuscules[["DNI"]])




Nombre de DNI amb lletres minúscules: 6
       DNI
89    None
104   None
114   None
142   None
281   None
1005   nan


**Conclusión: Solo encontró valores nulos. No hay ningún DNI con letra en minúscula en el registro.**

#» Hi ha valors obligatoris que falten o informació incoherent?


Hay multitud de elementos con información equivocada. Desde fechas con float, o formatos numéricos como fechas con datos imposibles, como dia 99 o mes 0. He detectado encabezados sin un formato adecuado, con espacios entre palabras en vez de guión bajo. Así como errores tipográficos. En la columna de Género también había datos no homogéneos.

#» Detectes formats incorrectes, errors de codificació o valors impossibles?


A continuación mostraré como he cambiado todos los nan a NA porque es la manera más moderna y con más adaptabilidad para tratar valores nulos. Es el estándar moderno recomendado porque respeta el tipo de dato original de la columna sin corromperlo.

Como he comentado previamente, ha habido errores de codificación tanto en los encabezados, como en los datos. El problema se ha corregido. Así como los datos imposibles en campos de fecha y demás.

**Homogeneización de datos: Columna Sexo**

In [85]:
# Sustituir 'D' por 'Dona', 'H' por 'Home' y n/d e 'A' por 'NC' directamente sobre df_sense_nuls
df_sense_nuls["Gènere"] = df_sense_nuls["Gènere"].replace({
    "D": "Dona",
    "H": "Home",
    "F": "Dona",
    "M": "Home",
    "NC": "n/c",
    "A": "n/c"
})

/tmp/ipykernel_665/810200084.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sense_nuls["Gènere"] = df_sense_nuls["Gènere"].replace({


**Homogeneización de datos: Contar cuantos salarios son con letras**

In [86]:

# 1. Filtramos solo las filas que contienen letras a la columna Salari mensual
salaris_lletres = df[
    df["Salari mensual"].astype(str).str.contains(r"[a-zA-Z]", regex=True)
]["Salari mensual"]

# 2. Contamos cuantas veces sale cada uno de estos textos
recompte = salaris_lletres.value_counts()

print(recompte)


Salari mensual
n/d               14
mil vuit-cents     9
dos mil            7
tres mil           6
1121 euros         3
1132 euros         2
2400 euros         1
3275 euros         1
1939 euros         1
1664 euros         1
1647 euros         1
1000 euros         1
1092 euros         1
1163 euros         1
889 euros          1
2488 euros         1
1987 euros         1
1058 euros         1
1244 euros         1
1019 euros         1
1324 euros         1
1849 euros         1
1405 euros         1
1457 euros         1
2648 euros         1
2028 euros         1
2600 euros         1
722 euros          1
1734 euros         1
1446 euros         1
1358 euros         1
1471 euros         1
1027 euros         1
2099 euros         1
1631 euros         1
1057 euros         1
1179 euros         1
1723 euros         1
1400 euros         1
1280 euros         1
1256 euros         1
2711 euros         1
881 euros          1
3061 euros         1
1250 euros         1
1668 euros         1
1880 euros         

**Salario: Limpieza de cualquier variación del símbolo de euro roto (amb â, ã, etc.), y puntos y comas eliminados**

In [87]:
df_sense_nuls["Salari mensual"] = (
    df_sense_nuls["Salari mensual"]
    .astype(str)
    .str.replace(r"[âã].*¬", "€", regex=True)
    .str.replace("euros", "€", regex=False)
    .str.replace(",", "", regex=False)
    .str.replace(".", "", regex=False)
)

/tmp/ipykernel_665/2312020816.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sense_nuls["Salari mensual"] = (


**Salario: Substitución letras por código**

In [88]:
# Creamos un diccionario con las equivalencias de texto a cifra numérica neta
reemplazos_salarios = {
    "mil vuit-cents": "1800€",
    "dos mil": "2000€",
    "tres mil": "3000€",
}

# Aplicamos el cambio a la columna Salari mensual directamente sobre df_sense_nuls
df_sense_nuls["Salari mensual"] = df_sense_nuls["Salari mensual"].replace(reemplazos_salarios)
df_sense_nuls["Salari mensual"] = df_sense_nuls["Salari mensual"].str.replace(" ", "", regex=False)

print(df_sense_nuls["Salari mensual"])

0       1469€
1       2718€
2       1358€
3       1478€
4       1284€
5       2069€
6       2349€
7       2855€
8       1338€
9       1450€
10       982€
11      2419€
12      3672€
13      1003€
14      1800€
15      1336€
16       982€
17      1385€
18      1751€
19       889€
20      1121€
21      1880€
22      1094€
23      1479€
24      1840€
26      2660€
27       751€
28      2628€
29      2400€
30      2564€
31       927€
32      2510€
33      1538€
34      3275€
35      1974€
36      1614€
37      1014€
38      2132€
39      2000€
41      1967€
42      1438€
43      1923€
44      1090€
45      1032€
46      2130€
47      1939€
48      1739€
49      1112€
50      1714€
51      1773€
52      1647€
53      1866€
54      1799€
55      1413€
56      1021€
57      1156€
58      1597€
59      1242€
60      2125€
61      3016€
62      1251€
63      1681€
64      2418€
65      1664€
66      1702€
67      1672€
68      2275€
69      1798€
70      1394€
71      1013€
72      1000€
73    

/tmp/ipykernel_665/489929084.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sense_nuls["Salari mensual"] = df_sense_nuls["Salari mensual"].replace(reemplazos_salarios)
/tmp/ipykernel_665/489929084.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sense_nuls["Salari mensual"] = df_sense_nuls["Salari mensual"].str.replace(" ", "", regex=False)


**Limpieza de datos: Columna Dia de Naixement, Mes de Naixement, Nombre de Fills y Any de Naixament**

In [89]:
df_sense_nuls['Dia de Naixement'] = df_sense_nuls['Dia de Naixement'].astype('Int64')
df_sense_nuls['Mes de Naixement'] = df_sense_nuls['Mes de Naixement'].astype('Int64')
df_sense_nuls['Any de Naixement'] = df_sense_nuls['Any de Naixement'].astype('Int64')
df_sense_nuls['Nombre_fills'] = df_sense_nuls['Nombre_fills'].astype('Int64')
df_sense_nuls['Km_anuals'] = df_sense_nuls['Km_anuals'].astype('Int64')

/tmp/ipykernel_665/3119000487.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sense_nuls['Dia de Naixement'] = df_sense_nuls['Dia de Naixement'].astype('Int64')
/tmp/ipykernel_665/3119000487.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sense_nuls['Mes de Naixement'] = df_sense_nuls['Mes de Naixement'].astype('Int64')
/tmp/ipykernel_665/3119000487.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value 

**Encabezados: Poner "_" a los encabezados y arreglar fallos tipográficos**

In [90]:
df_sense_nuls.columns = df_sense_nuls.columns.str.replace(' ', '_')
df_sense_nuls.columns.values[13] = 'Càrrec'

**Eliminar registros innecesarios (registros claramente inaprovechables)**

In [91]:
df_sense_nuls = df_sense_nuls.drop([1005, 1006])

**Fecha: Días con formato incorrecto**

In [92]:
import pandas as pd

# Convertir en nulos (<NA>) los dias que tienen más de 31 diass
df_sense_nuls.loc[df_sense_nuls['Dia_de_Naixement'] > 31, 'Dia_de_Naixement'] = pd.NA

# Convertir en nulos (<NA>) los meses que sean 0 o tengan més de 12 meses en el resigstro
df_sense_nuls.loc[
    (df_sense_nuls['Mes_de_Naixement'] == 0) |
    (df_sense_nuls['Mes_de_Naixement'] < 1) |
    (df_sense_nuls['Mes_de_Naixement'] > 12),
    'Mes_de_Naixement'
] = pd.NA

df_sense_nuls.loc[
    (df_sense_nuls['Dia_de_Naixement'] == 0) |
    (df_sense_nuls['Dia_de_Naixement'] < 1) |
    (df_sense_nuls['Dia_de_Naixement'] > 31),
    'Dia_de_Naixement'
] = pd.NA

**Solucionar la columna Kms_anuals con valores negativos o ilógicos**

In [93]:
df_sense_nuls.loc[df_sense_nuls['Km_anuals'] < 0, 'Km_anuals'] = pd.NA

**Limpieza de datos columna nacimiento: Años fuera de 2010 y 1950**

In [94]:
# 1. Convertir la columna a numérica (los textos inválidos o 'n/c' pasarán a NaN) - Se pasa a Numérico porque la columna es de string y por lo tanto NO deja hacer operaciones numericas como BETWEEN.
df_sense_nuls['Any_de_Naixement'] = pd.to_numeric(
    df_sense_nuls['Any_de_Naixement'])

# 2. Aplicar condición para convertir en pd.NA los valores fuera de rango (1950 - 2010)
df_sense_nuls.loc[
    df_sense_nuls['Any_de_Naixement'].between(1950, 2010) == False,
    'Any_de_Naixement',
] = pd.NA

# 3. Forzarlo a entero con soporte para nulos
df_sense_nuls['Any_de_Naixement'] = df_sense_nuls['Any_de_Naixement'].astype(
    'Int64'
)

**Nota: eliminar los registros de personas nacidas antes de 1950 y después de 2010 es un paso metódico para garantizar la coherencia y la fiabilidad del Dataset. Al tratarse de un dataset con variables laborales y salariales, excluir estos años de nacimiento "atípicos" y fuera de la edad de la población activa previene distorsiones estadísticas.**

**Limpieza de columna Temperatura: Datos incoherentes**

In [95]:
# 1. Asegurar de que la columna es numérica
df_sense_nuls['Temperatura_mitjana_ciutat'] = pd.to_numeric(
    df_sense_nuls['Temperatura_mitjana_ciutat'], errors='coerce'
)

# 2. Reemplazar por pd.NA aquellas temperaturas que sean menores o iguales a 0
df_sense_nuls.loc[
    df_sense_nuls['Temperatura_mitjana_ciutat'] <= 0,
    'Temperatura_mitjana_ciutat',
] = pd.NA

**Nota: He revisado que las temperaturas sean acorde a las ciudades y su clima habitual. Las que no, las he cambiado a NA (de momento).**

**Redundancia de datos: Eliminar columnas Fills y no Fills y corelacionar datos**

In [96]:
# 1. Asegurar formato numérico en Nombre_fills
df_sense_nuls['Nombre_fills'] = pd.to_numeric(
    df_sense_nuls['Nombre_fills'], errors='coerce'
)

# 2. Si la columna No_Fills existe, rellenar los nulos con 0 donde sea True
if 'No_Fills' in df_sense_nuls.columns:
    df_sense_nuls.loc[
        (df_sense_nuls['No_Fills'] == True)
        & (df_sense_nuls['Nombre_fills'].isna()),
        'Nombre_fills',
    ] = 0

# 3. Eliminar las columnas redundantes de forma segura del dataframe correcto
df_sense_nuls = df_sense_nuls.drop(
    columns=['Fills', 'No_Fills'], errors='ignore'
)

**Conversión de datos: Convertir todos los NaN y None en NA**

---



In [97]:
df_sense_nuls = df_sense_nuls.astype(str).replace('nan', '<NA>').replace('None', '<NA>')

# Data Frame con la limpieza de datos realizada (hasta este momento):

In [98]:
import pandas as pd

# Mostrar un número alto de filas para que no oculte ninguna (o pon el número total, ej. 1500)
pd.set_option('display.max_rows', 1500)

# Mostrar DataFrame
display(df_sense_nuls)

,Nom,Cognoms,DNI,País_d'origen,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Salari_mensual,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23,3,1958,Dona,1469€,Grup B,Cap de projecte,3,True,32108,25.0,10.1
1,Marc,Muñoz,48840994W,Espanya,Alacant,8,11,1960,Home,2718€,Grup C,Senior analyst,0,True,19496,10.4,18.7
2,Noa,Serra,14308421X,Espanya,Alacant,27,4,1961,Dona,1358€,Grup A,Tècnic IT,4,<NA>,<NA>,<NA>,16.7
3,Pol,Gil,58586340F,Espanya,Sevilla,13,10,1985,Home,1478€,Grup B,Data Analyst,2,False,<NA>,<NA>,18.3
4,David,Vila,82070937P,Espanya,Bilbao,29,11,1965,Dona,1284€,Grup B,Administratiu,<NA>,True,<NA>,11.8,13.1
5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1,8,1978,Dona,2069€,Grup B,Cap de projecte,0,<NA>,<NA>,<NA>,19.9
6,David,Gómez Vila,36998038E,Espanya,Palma,25,9,1971,Dona,2349€,Grup B,Cap de projecte,<NA>,True,6196,10.6,17.8
7,Carmen,López,47376585K,Espanya,Barcelona,8,9,1962,Home,2855€,Grup C,Responsable de vendes,4,True,20000,5.9,15.4
8,Nora,Andersen Pedersen,37289414J,Noruega,Stavanger,8,3,1963,Dona,1338€,Grup B,Analista junior,0,False,<NA>,<NA>,7.6
9,Alba,García,75181648S,Espanya,Las Palmas de Gran Canaria,<NA>,<NA>,<NA>,Home,1450€,Grup A,Senior analyst,4,False,<NA>,<NA>,20.1


#Exercici 3: Transformacions necessàries i preparació del dataset


**» Els valors categòrics estan normalitzats i són consistents?**

**Falta de normalización en el género (Gènere)**: En el Dataframe original las categorías no siguen un criterio unificado. En los datos originales existen abreviaturas y variantes distintas para referirse a lo mismo o introducir errores (por ejemplo, encontramos D, Dona, H, Home, M, NC, Donan/d o valores nulos y formatos extraños), lo que demuestra que los textos no estaban estandarizados.

**Inconsistencias de formato y codificación en los textos:** Debido a problemas de codificación, existen caracteres corrompidos en nombres, apellidos y ciudades (como Muñoz, Gómez, València, Díaz), rompiendo la consistencia.

**Valores atípicos y texto en campos numéricos:** Existen celdas con texto plano o descripciones donde debería haber datos normalizados (por ejemplo, sueldos escritos con letras como "mil vuit-cents" o formatos de moneda con caracteres extraños como "1,469 €").

Todo esto ha sido solucionado.

**» Els valors nuls són coherents amb la informació disponible?**

Sí son coherentes una vez procesados, ya que he aplicado una diferenciación lógica.

Es decir, he respetado el n/c como una respuesta válida de "no contesta" (n/c) marcada voluntariamente por el usuario, y he estandarizado el resto de ausencias a valores nulos técnicos nan como (<NA>), porque es la forma moderna de manejarlo hoy en día en herramientas como Python (Pandas).

**¿Por qué? Con el uso de NA:**

No alteramos los datos: Antes, poner un valor en una columna de números enteros (1, 2, 3) la transformaba a la fuerza en decimales (1.0, 2.0, NaN). El NA moderno mantiene los números enteros intactos.

Soporta cualquier tipo de información: Funciona exactamente igual en columnas de texto, fechas, números o valores booleanos, unificando todo bajo un solo estándar.





**» Hi ha redundància de variables que pugui generar problemes analítics?**

La información sobre los hijos: Teníamos columnas como Fills, No Fills y Nombre_fills mezcladas o repetidas midiendo prácticamente lo mismo (si tiene hijos o cuántos tiene). Por ese motivo he cruzado la información entre No Fills y Nombre_Fills y he eliminado Fills y No Fills por redundancia.

**» Les variables contextuals (com temperatura, salari) estan completes i preparades per a l'anàlisi?**

**Valores faltantes:** Tanto en las columnas de temperatura como en las de salario hay una gran cantidad de celdas vacías (NA) que impiden realizar análisis estadísticos directos sin un tratamiento previo de imputación

**Inconsistencias de formato y tipos de datos mezclados:** El salario no tiene un formato numérico homogéneo. Mientras unos valores son números limpios (ej. 1,469 €), otros están escritos total o parcialmente con texto (ej. 1358 euros, mil vuit-cents, dos mil) o contienen símbolos que impiden que el sistema los reconozca como números para calcular medias, medianas o desviaciones.

**Valores atípicos o erróneos**: En las variables meteorológicas o de temperatura aparecen valores ilógicos o extremos, que son claramente errores de registro o códigos de control que desvirtuarían cualquier análisis si no se limpian o convierten a nulos.

**» Normalitzar els valors categòrics per evitar ambigüitats (per exemple, valors que representen el mateix concepte però estan escrits de manera diferent).**

**Correcciones idiomáticas: Traducción al Catalán (idioma original del Dataset)**

In [99]:
# Diccionario de traducción para normalizar los valores categóricos de ciudades a catalán
traduccions_ciutats = {
    "Zaragoza": "Saragossa",
    "Milano": "Milà",
    "Genova": "Gènova",
    "Paris": "París",
    "A Coruña": "La Corunya",
    "Murcia": "Múrcia",
    "Lyon": "Lió",
    "Toulouse": "Tolosa",
    "Napoli": "Nàpols",
    "Venezia": "Venècia",
    "Bordeaux": "Bordeus",
    "Nice": "Niça",
    'München': 'Munic',
    "Firenze": "Florència",
    "Bologna": "Bolonya",
    "Torino": "Torí",
    "Marseille": "Marsella"
}

# Aplicar los cambios de forma homogénea en DataFrame df_sense_nuls
df_sense_nuls["Ciutat"] = df_sense_nuls["Ciutat"].replace(traduccions_ciutats)

In [100]:
# Diccionario de traducción y unificación para normalizar la columna País_d'origen
traduccions_paisos = {
    "Italia": "Itàlia",
}

# Aplicar los cambios de forma homogénea en el DataFrame
df_sense_nuls["País_d'origen"] = df_sense_nuls["País_d'origen"].replace(traduccions_paisos)

In [101]:
# Diccionario para unificar y traducir los cargos profesionales a catalán
traduccions_carregs = {
    "Senior analyst": "Analista sènior",
    "Data Analyst": "Analista de dades",
    "Analyst": "Analista",
    "Analista junior": "Analista júnior"
}

# Aplicar los cambios en la columna correspondiente del DataFrame
df_sense_nuls["Càrrec"] = df_sense_nuls["Càrrec"].replace(traduccions_carregs)

In [102]:
# Diccionario para unificar los cargos equivalentes al catalán
traduccions_carregs = {
    "Project lead": "Cap de projecte",
    "Senior analyst": "Analista sènior",
    "Data Analyst": "Analista de dades",
    "Data analyst": "Analista de dades",
    "Analyst": "Analista",
    "Admin.": "Administratiu/va",
    "Analista junior": "Analista júnior"
}

# Aplicar los cambios en la columna correspondiente del DataFrame
df_sense_nuls["Càrrec"] = df_sense_nuls["Càrrec"].replace(traduccions_carregs)

In [103]:
# Detectar filas donnde uno dels dos campps no tiene datos
incompleto = (df_sense_nuls['Consum_mitja_L_100km'].notna() & df_sense_nuls['Km_anuals'].isna()) | \
             (df_sense_nuls['Consum_mitja_L_100km'].isna() & df_sense_nuls['Km_anuals'].notna())

# 3. Assignar pd.NA
df_sense_nuls.loc[incompleto, ['Km_anuals', 'Consum_mitja_L_100km']] = pd.NA

In [104]:
# Comprobacón del DF en la versión actual
# display(df_sense_nuls)

**» Revisar la coherència dels valors nuls, assegurant que:
Les persones que tenen quilometratge anual i consum mitjà associats disposen de cotxe. En els casos en què consti consum mitjà però no quilometratge anual, caldrà decidir com resoldre aquesta incoherència de manera justificada.**

In [105]:
# Detectar filas donde hay incoherencia de datos
incompleta = (df_sense_nuls['Consum_mitja_L_100km'].notna() & df_sense_nuls['Km_anuals'].isna()) | \
             (df_sense_nuls['Consum_mitja_L_100km'].isna() & df_sense_nuls['Km_anuals'].notna())

# Poner <NA> tanto en el consumo como en los km anuales de estas filas
df_sense_nuls.loc[incompleta, ['Km_anuals', 'Consum_mitja_L_100km']] = pd.NA

In [106]:
display(df_sense_nuls)

,Nom,Cognoms,DNI,País_d'origen,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Salari_mensual,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23,3,1958,Dona,1469€,Grup B,Cap de projecte,3,True,32108,25.0,10.1
1,Marc,Muñoz,48840994W,Espanya,Alacant,8,11,1960,Home,2718€,Grup C,Analista sènior,0,True,19496,10.4,18.7
2,Noa,Serra,14308421X,Espanya,Alacant,27,4,1961,Dona,1358€,Grup A,Tècnic IT,4,<NA>,<NA>,<NA>,16.7
3,Pol,Gil,58586340F,Espanya,Sevilla,13,10,1985,Home,1478€,Grup B,Analista de dades,2,False,<NA>,<NA>,18.3
4,David,Vila,82070937P,Espanya,Bilbao,29,11,1965,Dona,1284€,Grup B,Administratiu,<NA>,True,<NA>,11.8,13.1
5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1,8,1978,Dona,2069€,Grup B,Cap de projecte,0,<NA>,<NA>,<NA>,19.9
6,David,Gómez Vila,36998038E,Espanya,Palma,25,9,1971,Dona,2349€,Grup B,Cap de projecte,<NA>,True,6196,10.6,17.8
7,Carmen,López,47376585K,Espanya,Barcelona,8,9,1962,Home,2855€,Grup C,Responsable de vendes,4,True,20000,5.9,15.4
8,Nora,Andersen Pedersen,37289414J,Noruega,Stavanger,8,3,1963,Dona,1338€,Grup B,Analista júnior,0,False,<NA>,<NA>,7.6
9,Alba,García,75181648S,Espanya,Las Palmas de Gran Canaria,<NA>,<NA>,<NA>,Home,1450€,Grup A,Analista sènior,4,False,<NA>,<NA>,20.1


**» Les persones que tenen un valor en el nombre de fills efectivament tenen fills.**

El uso de la condición con No_Fills permite deducir con total seguridad que si una persona indica explícitamente que no tiene hijos y el campo numérico está vacío, el valor real debe ser un 0 en lugar de dejarlo como un dato perdido.



**» Unificar la informació sobre fills en una única columna, ja que mantenir diverses columnes amb informació complementària no és òptim i pot generar problemes analítics (redundància o multicolinealitat).**

Eliminar las columnas redundantes (Fills y No_Fills) al final evita duplicidades de información y deja la estructura final optimizada exclusivamente con la variable numérica limpia.

**» Validar la temperatura mitjana. En cas que aquesta informació falti, hauràs d'imputar-la utilitzant un criteri raonable (per ciutat o per un nivell superior justificat).**

In [107]:
# 1. Assegurar formato numerico a la columna Temperatura_mitjana_ciutat
df_sense_nuls['Temperatura_mitjana_ciutat'] = pd.to_numeric(df_sense_nuls['Temperatura_mitjana_ciutat'], errors='coerce')

# 2. Imputar la temperatura faltante para ciutat y redondear directamente a 1 decimal
if 'Ciutat' in df_sense_nuls.columns:
    df_sense_nuls['Temperatura_mitjana_ciutat'] = df_sense_nuls.groupby('Ciutat')['Temperatura_mitjana_ciutat'].transform(
        lambda x: x.fillna(x.mean()).round(1)
    )

# 3. Assegurar que toda la columna quede a 1 decimal
df_sense_nuls['Temperatura_mitjana_ciutat'] = df_sense_nuls['Temperatura_mitjana_ciutat'].round(1)

Para resolver los valores nulos en la temperatura media, se han agrupado los registros por ciudad para calcular la media aritmética de cada localidad de forma independiente. A continuación, se han rellenado los NA con esa media local y se ha redondeado toda la columna a un solo decimal para mantener la coherencia y uniformidad de los datos.

In [108]:
# display(df_sense_nuls)

**» Revisar i imputar els salaris faltants, aplicant criteris analítics coherents (com ara càrrec, grup professional, ciutat, gènere o franja d'edat).**

Durante la revisión de los salarios, he optado por mantener los valores originales y respetar los registros indicados como no disponibles (n/d), evitando modificar datos sin conocimiento de la causa. No se ha considerado lo n/d como lo mismo que NA. Y no se ha encontrado ningún NA.

Asimismo, he considerado que aquellos salarios notablemente más bajos respecto al cargo o grupo professional —como el de algunos jefes de proyecto— pueden suceder de este modo por situaciones como trabajador media jornada o a circunstancias contractuales especiales propias, por lo que no se han clasificado automáticamente como errores.

Los salarios han sido normalizados y los cargos los he traducido a Catalán para tener una homogeanización de la información.

#Exercici 4: Primeres anàlisis descriptives


In [109]:
import numpy as np
import pandas as pd

# Netegem el símbol '€' de la columna de salaris i el convertim a numèric:
df_sense_nuls['Salari_mensual_num'] = (
    df_sense_nuls['Salari_mensual']
    .astype(str)
    .str.replace('€', '', regex=False)
    .str.strip()
)
df_sense_nuls['Salari_mensual_num'] = pd.to_numeric(
    df_sense_nuls['Salari_mensual_num'], errors='coerce'
)

# ==============================================================================
# 1. Construir taula resum de salaris per Gènere
# ==============================================================================
taula_resum_genere = (
    df_sense_nuls.groupby('Gènere')['Salari_mensual_num']
    .agg(
        Sou_Mitjà='mean', Sou_Mediana='median', Mínim='min', Màxim='max'
    )
    .sort_values(by='Sou_Mitjà', ascending=True)
)

# Neteja de la primera fila per posició si es queda un índex buit
if len(taula_resum_genere) > 3:
    taula_resum_genere = taula_resum_genere.iloc[1:]

# Eliminem el nom de l'índex per evitar desalineacions estètiques
taula_resum_genere.index.name = None
taula_resum_genere.columns.name = None

print('--- Exercici 4.1: Taula resum de salaris per Gènere ---\n')
display(
    taula_resum_genere.style.format('{:.2f}')
    .background_gradient(cmap='Blues', low=0.1, high=0.1)
    .set_table_styles([
        {
            'selector': 'th',
            'props': [
                ('background-color', '#2c3e50'),
                ('color', 'white'),
                ('font-family', 'sans-serif'),
                ('padding', '8px'),
            ],
        },
        {
            'selector': 'td',
            'props': [
                ('font-family', 'sans-serif'),
                ('padding', '8px'),
                ('text-align', 'center'),
            ],
        },
    ])
)

print('\n' + '=' * 60 + '\n')

# ==============================================================================
# 2. Taula creuada amb el salari mig per Gènere i País d'origen
# ==============================================================================
taula_creuada = df_sense_nuls.pivot_table(
    values='Salari_mensual_num',
    index='Gènere',
    columns="País_d'origen",
    aggfunc='mean',
    margins=True,
    margins_name='Total',
)

# Neteja de la primera fila si s'hagués colat un índex no desitjat
if len(taula_creuada) > 4:
    taula_creuada = taula_creuada.iloc[1:]

# Eliminar els títols d'índex i columnes per evitar desalineacions (Gènere / País_d'origen flotants)
taula_creuada.index.name = None
taula_creuada.columns.name = None

print(
    '--- Exercici 4.2: Taula creuada (Salari mig per Gènere i País d\'origen) ---\n'
)
display(
    taula_creuada.style.format('{:.2f}', na_rep='-')
    .background_gradient(cmap='YlGnBu', low=0.1, high=0.1)
    .set_table_styles([
        {
            'selector': 'th',
            'props': [
                ('background-color', '#2c3e50'),
                ('color', 'white'),
                ('font-family', 'sans-serif'),
                ('padding', '8px'),
            ],
        },
        {
            'selector': 'td',
            'props': [
                ('font-family', 'sans-serif'),
                ('padding', '8px'),
                ('text-align', 'center'),
            ],
        },
    ])
)

--- Exercici 4.1: Taula resum de salaris per Gènere ---



,Sou_Mitjà,Sou_Mediana,Mínim,Màxim
Dona,1485.69,1358.00,701.00,3564.00
n/c,1621.96,1256.00,673.00,3600.00
Home,1694.72,1509.00,726.00,3672.00




--- Exercici 4.2: Taula creuada (Salari mig per Gènere i País d'origen) ---



,Alemanya,Espanya,França,Itàlia,Noruega,Total
Dona,1429.91,1482.77,1411.48,1473.64,1779.75,1485.69
Home,2103.82,1699.51,1672.53,1662.35,1479.00,1694.72
n/c,1634.00,1628.33,1229.00,2047.40,1237.00,1621.96
Total,1602.50,1566.84,1501.01,1567.12,1641.23,1566.30


**Conclusiones:**

**Disparidad salarial:** Globalmente, los hombres tienen un sueldo medio superior (1.694,72€) en comparación con las mujeres (1.485,69€), tal y como podemos ver en el gráfico de "Exercici 1".

**Valores extremos:** Tanto el salario mínimo absoluto como el máximo absoluto se encuentran en el colectivo masculino, aunque las mujeres registran un rango muy similar.

**Excepciones**: A pesar de la tendencia general, se observan excepciones importantes según el país de origen. En Noruega, las mujeres presentan un salario medio significativamente más alto (1.779,75€) que los hombres (1.479,00 €) acorde los datos que tenemos disponibles.

...Por otra parte, Alemania destaca por registrar la media salarial más elevada en hombres (2.103,82€), siendo uno de los salarios más altos de toda la tabla cruzada.

En cuanto a los N/C, podemos ver que en la mayoría de los casos se encuentran entre medio de los valores obtenidos entre hombres y mujeres, salvo en Italia.

**Nota**: Los NA han sido descartados del gráfico para asegurar que las métricas y conclusiones se basen exclusivamente en datos reales y válidos.

#Nivell 2 - Anàlisi amb dades enriquides i estructures múltiples

#Exercici 1: Enriquiment mitjançant datasets externs

**Merge entre el DataFrame anterior y el CSV de Ciutats_Context**

In [110]:
import pandas as pd

# Definir la ruta de carpeta en el Drive
ruta_carpeta = '/content/drive/My Drive/Pandas_Sprint_8/'

# Cargar el archivo de ciudades desde el Drive
df_ciutats = pd.read_csv(
    ruta_carpeta + 'ciutats_context.csv', encoding='latin1'
)


# --- 1. NETEJA I HOMOLOGACIÓ DE CIUTATS ---
# Convertir a string i netejar possibles  invertides de text tipus \u00f1
df_ciutats['Ciutat'] = (
    df_ciutats['Ciutat']
    .astype(str)
    .str.replace(r'\\u00f1', 'ñ', regex=True)
    .str.replace(r'\\u00e0', 'à', regex=True)
    .str.replace(r'\\u00e8', 'è', regex=True)
    .str.replace(r'\\u00ec', 'ì', regex=True)
    .str.replace(r'\\u00f2', 'ò', regex=True)
    .str.replace(r'\\u00f9', 'ù', regex=True)
)

# Diccionario
reemplazos_ciutats = {
    'Zaragoza': 'Saragossa',
    'Milano': 'Milà',
    'Genova': 'Gènova',
    'Paris': 'París',
    'A Coruña': 'La Corunya',  # Por si ya tiene la ñ
    'A Coru\u00f1a': 'La Corunya',  # Por si viene con la barra invertida literal
    'A Coru\\u00f1a': 'La Corunya',  # Por si tiene doble barra
    'Murcia': 'Múrcia',
    'Lyon': 'Lió',
    'Toulouse': 'Tolosa',
    'Napoli': 'Nàpols',
    'Venezia': 'Venècia',
    'Bordeaux': 'Bordeus',
    'Nice': 'Niça',
    'Munchen': 'Munic',
    'Firenze': 'Florència',
    'Bologna': 'Bolonya',
    'Torino': 'Torí',
    'Marseille': 'Marsella',
}

df_ciutats['Ciutat'] = df_ciutats['Ciutat'].replace(reemplazos_ciutats)

# Fusionar (merge) dataframe limpio con el de ciudades del Drive
df_enriquit = pd.merge(df_sense_nuls, df_ciutats, on='Ciutat', how='left')

# Eliminar columnas duplicadas
if 'País' in df_enriquit.columns:
    df_enriquit = df_enriquit.drop(columns=['País'])
if 'Salari_mensual_num' in df_enriquit.columns:
    df_enriquit = df_enriquit.drop(columns=['Salari_mensual_num'])

# Pasar los NaN a NA
df_enriquit = df_enriquit.convert_dtypes()

# Comprobar el resultado
print(f'\nFilas totales del dataset enriquecido: {df_enriquit.shape[0]}')
pd.set_option('display.max_columns', None)
display(df_enriquit.head(30))



Filas totales del dataset enriquecido: 997


,Nom,Cognoms,DNI,País_d'origen,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Salari_mensual,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat,Cost_vida_index,Transport_public_score,Clima_tipus
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23,3,1958,Dona,1469€,Grup B,Cap de projecte,3,True,32108,25.0,10.1,64,6.7,Continental
1,Marc,Muñoz,48840994W,Espanya,Alacant,8,11,1960,Home,2718€,Grup C,Analista sènior,0,True,19496,10.4,18.7,68,7.0,Mediterrani
2,Noa,Serra,14308421X,Espanya,Alacant,27,4,1961,Dona,1358€,Grup A,Tècnic IT,4,<NA>,<NA>,<NA>,16.7,68,7.0,Mediterrani
3,Pol,Gil,58586340F,Espanya,Sevilla,13,10,1985,Home,1478€,Grup B,Analista de dades,2,False,<NA>,<NA>,18.3,65,7.0,Mediterrani
4,David,Vila,82070937P,Espanya,Bilbao,29,11,1965,Dona,1284€,Grup B,Administratiu,<NA>,True,<NA>,11.8,13.1,72,7.8,Oceànic
5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1,8,1978,Dona,2069€,Grup B,Cap de projecte,0,<NA>,<NA>,<NA>,19.9,62,6.8,Subtropical
6,David,Gómez Vila,36998038E,Espanya,Palma,25,9,1971,Dona,2349€,Grup B,Cap de projecte,<NA>,True,6196,10.6,17.8,73,7.2,Mediterrani
7,Carmen,López,47376585K,Espanya,Barcelona,8,9,1962,Home,2855€,Grup C,Responsable de vendes,4,True,20000,5.9,15.4,78,8.5,Mediterrani
8,Nora,Andersen Pedersen,37289414J,Noruega,Stavanger,8,3,1963,Dona,1338€,Grup B,Analista júnior,0,False,<NA>,<NA>,7.6,<NA>,<NA>,<NA>
9,Alba,García,75181648S,Espanya,Las Palmas de Gran Canaria,<NA>,<NA>,<NA>,Home,1450€,Grup A,Analista sènior,4,False,<NA>,<NA>,20.1,62,6.8,Subtropical



**» Integrar el dataset principal amb:**

El dataset contextual de ciutats.

In [111]:
import pandas as pd

# 1. Cargamos el dataset de ciudades
df_ciudades = pd.read_csv("/content/drive/MyDrive/Pandas_Sprint_8/ciutats_context.csv", encoding='latin1')

# Limpieza de caracteres codificados
df_ciudades['Ciutat'] = (
    df_ciudades['Ciutat']
    .astype(str)
    .str.replace(r'\\u00f1', 'ñ', regex=True)
    .str.replace(r'\\u00e0', 'à', regex=True)
    .str.replace(r'\\u00e8', 'è', regex=True)
    .str.replace(r'\\u00ec', 'ì', regex=True)
    .str.replace(r'\\u00f2', 'ò', regex=True)
    .str.replace(r'\\u00f9', 'ù', regex=True)
)

# Diccionario completo
reemplazos_ciutats = {
    'Zaragoza': 'Saragossa',
    'Milano': 'Milà',
    'Genova': 'Gènova',
    'Paris': 'París',
    'A Coruña': 'La Corunya',  # Por si ya tiene la ñ
    'A Coru\u00f1a': 'La Corunya',  # Por si viene con la barra invertida literal
    'A Coru\\u00f1a': 'La Corunya',  # Por si tiene doble barra
    'Murcia': 'Múrcia',
    'Lyon': 'Lió',
    'Toulouse': 'Tolosa',
    'Napoli': 'Nàpols',
    'Venezia': 'Venècia',
    'Bordeaux': 'Bordeus',
    'Nice': 'Niça',
    'Munchen': 'Munic',
    'Firenze': 'Florència',
    'Bologna': 'Bolonya',
    'Torino': 'Torí',
    'Marseille': 'Marsella',
}

# 2. Aplicamos el diccionario correcto (reemplazos_ciutats) y limpieza de espacios
df_ciudades["Ciutat"] = df_ciudades["Ciutat"].replace(reemplazos_ciutats).str.strip()

# 3. Eliminamos las columnas antiguas del DF principal
columnas_a_borrar = ["País_d'origen", 'País', 'Cost_vida_index', 'Transport_public_score', 'Clima_tipus']
df_enriquit = df_enriquit.drop(columns=columnas_a_borrar, errors='ignore')

# 4. Aseguramos que la columna de unión en el dataframe principal también esté limpia
df_enriquit["Ciutat"] = df_enriquit["Ciutat"].str.strip()

# 5. Hacemos el merge
df_enriquit = pd.merge(df_enriquit, df_ciudades, on='Ciutat', how='left')

# 6. Mostramos el resultado
df_enriquit.head(5)

,Nom,Cognoms,DNI,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Salari_mensual,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat,País,Cost_vida_index,Transport_public_score,Clima_tipus
0,Joana,Gil Navarro,66722344X,Valladolid,23,3,1958,Dona,1469€,Grup B,Cap de projecte,3,True,32108,25.0,10.1,Espanya,64.0,6.7,Continental
1,Marc,Muñoz,48840994W,Alacant,8,11,1960,Home,2718€,Grup C,Analista sènior,0,True,19496,10.4,18.7,Espanya,68.0,7.0,Mediterrani
2,Noa,Serra,14308421X,Alacant,27,4,1961,Dona,1358€,Grup A,Tècnic IT,4,<NA>,<NA>,<NA>,16.7,Espanya,68.0,7.0,Mediterrani
3,Pol,Gil,58586340F,Sevilla,13,10,1985,Home,1478€,Grup B,Analista de dades,2,False,<NA>,<NA>,18.3,Espanya,65.0,7.0,Mediterrani
4,David,Vila,82070937P,Bilbao,29,11,1965,Dona,1284€,Grup B,Administratiu,<NA>,True,<NA>,11.8,13.1,Espanya,72.0,7.8,Oceànic


**» Integrar la política salarial definida en format JSON**

In [112]:
import json
import pandas as pd

# Cargamos el archivo JSON de la política salarial
with open('/content/drive/MyDrive/Pandas_Sprint_8/politica_salarial.json', 'r', encoding='utf-8') as f:
    politica = json.load(f)

# Asignamos un porcentaje de incremento salarial a cada empleado en una nueva columna (Inc_Grupo) buscando el valor correspondiente a su grupo profesional en el archivo JSON.
df_enriquit['Inc_Grupo'] = df_enriquit['Grup_Professional'].map(politica['base_increment_by_group'])

# Visualizamos el resultado con las nuevas columnas aplicadas
df_enriquit.head(10)

,Nom,Cognoms,DNI,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Salari_mensual,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat,País,Cost_vida_index,Transport_public_score,Clima_tipus,Inc_Grupo
0,Joana,Gil Navarro,66722344X,Valladolid,23,3,1958,Dona,1469€,Grup B,Cap de projecte,3,True,32108,25.0,10.1,Espanya,64.0,6.7,Continental,3.5
1,Marc,Muñoz,48840994W,Alacant,8,11,1960,Home,2718€,Grup C,Analista sènior,0,True,19496,10.4,18.7,Espanya,68.0,7.0,Mediterrani,2.0
2,Noa,Serra,14308421X,Alacant,27,4,1961,Dona,1358€,Grup A,Tècnic IT,4,<NA>,<NA>,<NA>,16.7,Espanya,68.0,7.0,Mediterrani,5.0
3,Pol,Gil,58586340F,Sevilla,13,10,1985,Home,1478€,Grup B,Analista de dades,2,False,<NA>,<NA>,18.3,Espanya,65.0,7.0,Mediterrani,3.5
4,David,Vila,82070937P,Bilbao,29,11,1965,Dona,1284€,Grup B,Administratiu,<NA>,True,<NA>,11.8,13.1,Espanya,72.0,7.8,Oceànic,3.5
5,Alexia,Puig,25014631B,Las Palmas de Gran Canaria,1,8,1978,Dona,2069€,Grup B,Cap de projecte,0,<NA>,<NA>,<NA>,19.9,Espanya,62.0,6.8,Subtropical,3.5
6,David,Gómez Vila,36998038E,Palma,25,9,1971,Dona,2349€,Grup B,Cap de projecte,<NA>,True,6196,10.6,17.8,Espanya,73.0,7.2,Mediterrani,3.5
7,Carmen,López,47376585K,Barcelona,8,9,1962,Home,2855€,Grup C,Responsable de vendes,4,True,20000,5.9,15.4,Espanya,78.0,8.5,Mediterrani,2.0
8,Nora,Andersen Pedersen,37289414J,Stavanger,8,3,1963,Dona,1338€,Grup B,Analista júnior,0,False,<NA>,<NA>,7.6,NaN,NaN,NaN,NaN,3.5
9,Alba,García,75181648S,Las Palmas de Gran Canaria,<NA>,<NA>,<NA>,Home,1450€,Grup A,Analista sènior,4,False,<NA>,<NA>,20.1,Espanya,62.0,6.8,Subtropical,5.0


**» Calcular un increment salarial total combinant:**

Grup professional.

Situació familiar.

Franja d'edat.

Cost de vida de la ciutat.


In [113]:
import pandas as pd
import json

# 1. Cargar la política desde el archivo JSON
with open('/content/drive/MyDrive/Pandas_Sprint_8/politica_salarial.json', 'r', encoding='utf-8') as f:
    politica = json.load(f)

# 2. Preparar los datos necesarios para el cálculo
edad = 2026 - pd.to_numeric(df_enriquit['Any_de_Naixement'], errors='coerce')
franja = pd.cut(edad, bins=[0, 29, 44, 59, 120], labels=['18-29', '30-44', '45-59', '60+'])
coste = pd.cut(df_enriquit['Cost_vida_index'].fillna(df_enriquit['Cost_vida_index'].mean()), bins=[0, 65, 75, 100], labels=['low', 'medium', 'high'])
fills_num = pd.to_numeric(df_enriquit['Nombre_fills'], errors='coerce').fillna(0)

# 3. Calcular el incremento total sumando cada factor convertido explícitamente a float
df_enriquit['Increment_Salari'] = (
    df_enriquit['Grup_Professional'].map(politica['base_increment_by_group']).fillna(0).astype(float) +
    (fills_num > 0).map({True: politica['children_adjustment']['has_children'], False: politica['children_adjustment']['no_children']}).astype(float) +
    franja.map(politica['age_adjustment']).fillna(0).astype(float) +
    coste.map(politica['cost_of_living_adjustment']).fillna(0).astype(float)
)

# 4. Aplicar el tope máximo (cap) definido en las reglas del JSON
max_tope = politica['rules']['max_total_increment_percent']
df_enriquit['Increment_Salari'] = df_enriquit['Increment_Salari'].clip(upper=max_tope)

# 5. Formatear la columna para que muestre el símbolo '%' al lado
df_enriquit['Increment_Salari'] = df_enriquit['Increment_Salari'].apply(lambda x: f"{x}%")

# 6. Visualizar las primeras 10 filas de todo el DataFrame
df_enriquit.head(40)

,Nom,Cognoms,DNI,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Salari_mensual,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat,País,Cost_vida_index,Transport_public_score,Clima_tipus,Inc_Grupo,Increment_Salari
0,Joana,Gil Navarro,66722344X,Valladolid,23,3,1958,Dona,1469€,Grup B,Cap de projecte,3,True,32108,25.0,10.1,Espanya,64.0,6.7,Continental,3.5,6.5%
1,Marc,Muñoz,48840994W,Alacant,8,11,1960,Home,2718€,Grup C,Analista sènior,0,True,19496,10.4,18.7,Espanya,68.0,7.0,Mediterrani,2.0,4.5%
2,Noa,Serra,14308421X,Alacant,27,4,1961,Dona,1358€,Grup A,Tècnic IT,4,<NA>,<NA>,<NA>,16.7,Espanya,68.0,7.0,Mediterrani,5.0,9.0%
3,Pol,Gil,58586340F,Sevilla,13,10,1985,Home,1478€,Grup B,Analista de dades,2,False,<NA>,<NA>,18.3,Espanya,65.0,7.0,Mediterrani,3.5,6.0%
4,David,Vila,82070937P,Bilbao,29,11,1965,Dona,1284€,Grup B,Administratiu,<NA>,True,<NA>,11.8,13.1,Espanya,72.0,7.8,Oceànic,3.5,6.0%
5,Alexia,Puig,25014631B,Las Palmas de Gran Canaria,1,8,1978,Dona,2069€,Grup B,Cap de projecte,0,<NA>,<NA>,<NA>,19.9,Espanya,62.0,6.8,Subtropical,3.5,5.5%
6,David,Gómez Vila,36998038E,Palma,25,9,1971,Dona,2349€,Grup B,Cap de projecte,<NA>,True,6196,10.6,17.8,Espanya,73.0,7.2,Mediterrani,3.5,6.5%
7,Carmen,López,47376585K,Barcelona,8,9,1962,Home,2855€,Grup C,Responsable de vendes,4,True,20000,5.9,15.4,Espanya,78.0,8.5,Mediterrani,2.0,7.0%
8,Nora,Andersen Pedersen,37289414J,Stavanger,8,3,1963,Dona,1338€,Grup B,Analista júnior,0,False,<NA>,<NA>,7.6,NaN,NaN,NaN,NaN,3.5,6.0%
9,Alba,García,75181648S,Las Palmas de Gran Canaria,<NA>,<NA>,<NA>,Home,1450€,Grup A,Analista sènior,4,False,<NA>,<NA>,20.1,Espanya,62.0,6.8,Subtropical,5.0,6.5%


**Añadimos columna nueva con el nuevo salario calculado**

In [114]:
# 1. Limpiar el salario mensual (quitar el símbolo '€' y pasarlo a número)
salario_num = df_enriquit['Salari_mensual'].astype(str).str.replace('€', '').str.strip()
salario_num = pd.to_numeric(salario_num, errors='coerce')

# 2. Calcular el nuevo salario aplicando el incremento porcentual obtenido previamente
df_enriquit['Salari_Nou'] = salario_num * (1 + df_enriquit['Increment_Salari'].str.replace('%', '').astype(float) / 100)

# 3. Formatear el resultado final con el símbolo '€' (redondeado a 2 decimales)
df_enriquit['Salari_Nou'] = df_enriquit['Salari_Nou'].round(2).apply(lambda x: f"{x}€" if pd.notnull(x) else x)

# 4. Ver el resultado
df_enriquit[['Nom','Ciutat', 'Salari_mensual', 'Increment_Salari', 'Salari_Nou']].head(10)

,Nom,Ciutat,Salari_mensual,Increment_Salari,Salari_Nou
0,Joana,Valladolid,1469€,6.5%,1564.48€
1,Marc,Alacant,2718€,4.5%,2840.31€
2,Noa,Alacant,1358€,9.0%,1480.22€
3,Pol,Sevilla,1478€,6.0%,1566.68€
4,David,Bilbao,1284€,6.0%,1361.04€
5,Alexia,Las Palmas de Gran Canaria,2069€,5.5%,2182.8€
6,David,Palma,2349€,6.5%,2501.68€
7,Carmen,Barcelona,2855€,7.0%,3054.85€
8,Nora,Stavanger,1338€,6.0%,1418.28€
9,Alba,Las Palmas de Gran Canaria,1450€,6.5%,1544.25€


**» Identificar quins registres no poden rebre algun dels ajustos i explicar-ne el motiu.**



In [115]:
# Comprobamos qué registros tienen valores nulos en las columnas clave que alimentan los ajustes
registros_con_faltas = df_enriquit[
    df_enriquit['Any_de_Naixement'].isna() |
    df_enriquit['Nombre_fills'].isna() |
    df_enriquit['Cost_vida_index'].isna()
]

# Mostramos cuántos son y cuáles faltan
print(f"Total de registros afectados: {len(registros_con_faltas)}")
registros_con_faltas[['Nom', 'Ciutat','Any_de_Naixement', 'Nombre_fills', 'Cost_vida_index']].head(20)

Total de registros afectados: 131


,Nom,Ciutat,Any_de_Naixement,Nombre_fills,Cost_vida_index
8,Nora,Stavanger,1963,0,NaN
21,Sergio,Màlaga,1972,4,NaN
29,Chloé,Tolosa,1973,2,NaN
37,Hugo,Màlaga,1961,0,NaN
53,Marina,Màlaga,1988,0,NaN
64,Jon,Trondheim,1978,<NA>,NaN
74,lise,Niça,1960,2,NaN
100,Chiara,Bolonya,1969,3,NaN
102,Arthur,Niça,1973,<NA>,NaN
116,Maja,Stavanger,1975,4,NaN


Tras realizar el merge, se han identificado registros con valores NA en las columnas añadidas. El motivo es que el dataset de referencia (ciutats_context.csv) no contiene todas las ciudades presentes en el DataFrame original.

Por ejemplo, ciudades como Málaga no aparecen en el archivo de contexto, por lo que no es posible asignarles los valores correspondientes de Cost_vida_index, Transport_public_score ni Clima_tipus.

En consecuencia, el merge mantiene deja los campos de contexto en NA al no encontrar una clave de unión válida en la tabla de referencia.

Para solucionar este problema en el futuro, sería necesario o bien actualizar el archivo ciutats_context.csv para incluir las ciudades faltantes, como Málaga o Stavanger.

**» Decidir, de manera raonada, com tractar aquests casos en el càlcul del salari ajustat, justificando l'impacte analític de la decisió presa.**

Dado que eliminar 131 registros representa una pérdida significativa de información que podría comprometer la representatividad de la muestra, se opta por imputar el valor medio nacional (.fillna(df_enriquit['Cost_vida_index'].mean())) a las ciudades sin contexto del CSV del Dataset. Esto permite mantener la integridad de los datos salariales originales sin descartar observaciones.

El impacto analítico implica una reducción de la precisión en el ajuste salarial para esas 131 ciudades. Por lo que el plan B sería eliminar estos registros si se considerara oportuno.

#Aspectes a revisar i analitzar



**» Les diferents fonts de dades són coherents i compatibles entre si per poder
calcular un salari ajustat?**

Sí, las fuentes (el dataset principal, el de ciudades y el JSON de política salarial) se han podido cruzar a través de claves comunes como la ciudad y el grupo profesional, aunque se ha requerido una limpieza previa de formatos (como eliminar símbolos de moneda y convertir tipos de texto a numéricos). Luego se han eliminado columnas redundantes (Repetidas).

**» La informació contextual incorporada (cost de vida, polítiques salarials) és suficient per aplicar increments de manera consistent?**

Sí, el diccionario JSON proporciona una estructura clara de reglas (por grupos, hijos, edad, coste de vida y transporte) que se aplica a todo el conjunto de datos.

**» Hi ha buits d'informació que impedeixin calcular correctament l'increment salarial en alguns registres?**

Sí, y tanto. Se ha detectado un total de 131 registros que presentan una ciudad de origen que no aparece en el CSV ciutats_context.csv, por lo tanto, al no haber coincidencia con los datos, directamente aparece como NA.

Principalmente son datos que tienen CIUDADES que NO aparecen en el CSV. Como por ejemplo Málaga.

**» És necessari prendre decisions analítiques quan la informació contextual és parcial o absent?**

Sí, porque no siempre tenemos toda la información necesaria. En este caso, he decidido mantener los NA para no perder esos registros y tenerlos disponibles para otros análisis.

**» Com poden aquestes decisions influir en el càlcul del salari ajustat i en la seva interpretació posterior?**




Decidir cómo gestionar los datos ausentes o incorrectos (como rellenar los valores nulos en los hijos con un 0, asignar una edad por defecto si falta el año de nacimiento) evita que el código falle por errores de tipo o claves inexistentes. Esto permite procesar toda la muestra de la plantilla de forma fluida y obtener el resultado numérico del incremento salarial.

#Exercici 2: Reestructuració del dataset


**» Existeixen variables que representen conceptes similars o relacionats, però estan distribuïdes en diferents columnes?**

Sí.

En el dataset original en formato ancho, encontramos variables de la misma categoría separadas en distintas columnas (como Salari_base y Salari_Ajustat), así como otros de tema urbano (como Cost_vida_index, Transport_public_score o Percentatge_Increment). Y esto para hacer análisis de datos, es un poco más complicado.

Por lo que a resumidas cuentas...

El formato ancho es más cómodo para consultar los datos visualmente, mientras que el formato largo facilita operaciones como agrupar, filtrar y comparar métricas.

**»El format actual del dataset facilita o dificulta: comparacions entre magnituds i agregacions per grups.**

Lo dificulta.

Al estar cada dato en su propia columna con escalas totalmente diferentes (por ejemplo, un salario de miles de euros frente a una puntuación de transporte de 1 a 10), resulta complejo realizar comparaciones directas.

Obliga a calcular columna por columna de forma manual y repetitiva, en lugar de permitir una agrupación global y automatizada de todas las métricas a la vez.

**» Quin impacte té la forma del dataset (ample vs llarg) en la lectura i interpretació de les dades?**

Formato ancho (Wide): Como cité anteriormente, facilita la lectura visual humana, pero dificulta el análisis automatizado y la creación de gráficos.

Formato largo (Long): Multiplica el número de filas, pero ofrece una interpretación y estructura mucho más adecuados para el análisis de datos. Permite tratar cualquier métrica, facilitando el análisis y las visualizaciones de datos.

**» Reestructurar el dataset utilitzant melt() per convertir-lo a format llarg, definint prèviament: les columnes que actuen com a identificadores (per exemple: persona, ciutat, grup professional, gènere, etc.). Les columnes que passen a representar valors mesurats (per exemple: salaris, mobilitat, variables contextuals).**

In [116]:
import pandas as pd

# 1. Columnes identificadores (quitamos País_d'origen si ya no está)
columnes_id = [
    'Nom', 'Cognoms', 'DNI', 'Ciutat',
    'Dia_de_Naixement', 'Mes_de_Naixement', 'Any_de_Naixement',
    'Gènere', 'Grup_Professional', 'Càrrec', 'Nombre_fills',
    'Te_cotxe', 'Clima_tipus'
]

# 2. Columnes que passen a representar valors mesurats (con los nombres reales)
columnes_a_fusionar = [
    'Salari_mensual', 'Km_anuals', 'Consum_mitja_L_100km',
    'Temperatura_mitjana_ciutat', 'Cost_vida_index',
    'Transport_public_score',
    'Increment_Salari', 'Salari_Nou'
]

# 3. Aplicar pd.melt()
df_llarg = pd.melt(
    df_enriquit,
    id_vars=columnes_id,
    value_vars=columnes_a_fusionar,
    var_name='Metrica_Tipus',
    value_name='Valor_Numeric'
)

# 4. Ordenar per DNI perquè les files de cada persona vagin juntes
df_llarg_ordenat = df_llarg.sort_values(by=['DNI', 'Metrica_Tipus']).reset_index(drop=True)

# 5. Mostrar dimensions i la mostra del dataset ORDENAT
display(df_llarg_ordenat.head(20))

,Nom,Cognoms,DNI,Ciutat,Dia_de_Naixement,Mes_de_Naixement,Any_de_Naixement,Gènere,Grup_Professional,Càrrec,Nombre_fills,Te_cotxe,Clima_tipus,Metrica_Tipus,Valor_Numeric
0,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Consum_mitja_L_100km,<NA>
1,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Cost_vida_index,62.0
2,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Increment_Salari,6.0%
3,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Km_anuals,<NA>
4,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Salari_Nou,1002.76€
5,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Salari_mensual,946€
6,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Temperatura_mitjana_ciutat,22.7
7,Claudia,Pujol Torres,10161088K,Las Palmas de Gran Canaria,27,9,1996,Home,Grup A,Tècnic IT,<NA>,False,Subtropical,Transport_public_score,6.8
8,Oriol,Torres Vila,10443906N,Tarragona,8,9,1962,Dona,Grup B,Responsable de vendes,1,False,Mediterrani,Consum_mitja_L_100km,<NA>
9,Oriol,Torres Vila,10443906N,Tarragona,8,9,1962,Dona,Grup B,Responsable de vendes,1,False,Mediterrani,Cost_vida_index,67.0


**» Analitzar per què el format llarg és més adequat per: realitzar comparacions entre diferents tipus de variables. Calcular estadístics agregats.**

**Realizar comparaciones entre diferentes tipus de variables**

Al tener todas las métricas numéricas unificadas en una sola columna de valores y otra de tipos, las herramientas de visualización pueden utilizar directamente la columna de tipos para crear gráficos de barras, etc de manera automática.

Facilita el uso de filtros rápidos y operaciones para comparar distintas variables entre sí sin tener que referenciar manualmente columnas individualmente.

**Calcular estadístics agregats**

Con el formato largo puedo usar groupby de forma más sencilla para calcular medias, sumas u otros estadísticos según la métrica.

Y así evitar tener que calcular los estadísticos columna por columna por separado, como se tendría que hacer en formato ancho, haciéndolo de manera más eficiente.

#Exercici 3: Creació de variables derivades (Feature Engineering)

**1. Ràtio salari / cost de vida**

**Fórmula principal:** Ratio Salari Cost de Vida = Salari Mensual / Cost de Vida Index

In [117]:
# Aquí limpiamos el símbolo de euro, eliminamos los puntos de miles y convertimos a entero
df_enriquit['Salari_mensual'] = (
    df_enriquit['Salari_mensual']
    .astype(str)
    .str.replace('€', '', regex=False)
    .str.replace('.', '', regex=False)
    .str.replace(',', '.', regex=False)
)
df_enriquit['Salari_mensual'] = pd.to_numeric(df_enriquit['Salari_mensual'], errors='coerce').astype('Int64')

# Calcular la ratio (redondeada a 2 decimales)
df_enriquit['Ratio_Salari_CostVida'] = (df_enriquit['Salari_mensual'] / df_enriquit['Cost_vida_index']).round(2)

# Filtrar para eliminar los registros incompletos (donde aparezcan valores NA)
df_filtrat_ratio = df_enriquit.dropna(subset=['Ratio_Salari_CostVida'])

# Muestra
display(df_filtrat_ratio[['Nom', 'Cognoms', 'Salari_mensual', 'Cost_vida_index', 'Ratio_Salari_CostVida']].head(20))

,Nom,Cognoms,Salari_mensual,Cost_vida_index,Ratio_Salari_CostVida
0,Joana,Gil Navarro,1469,64.0,22.95
1,Marc,Muñoz,2718,68.0,39.97
2,Noa,Serra,1358,68.0,19.97
3,Pol,Gil,1478,65.0,22.74
4,David,Vila,1284,72.0,17.83
5,Alexia,Puig,2069,62.0,33.37
6,David,Gómez Vila,2349,73.0,32.18
7,Carmen,López,2855,78.0,36.6
9,Alba,García,1450,62.0,23.39
10,Joana,Torres Díaz,982,64.0,15.34


**La seva utilitat analítica.**

Con el ratio de Salario / Coste de vida obtenemos el índex relativo de poder adquisitivo. Por lo que, podemos saber el nivel de vida de un individuo en base a su remuneración y su vivienda actual (coste de vida).



**Les decisions preses.**

**Limpieza y estandarización de cadenas monetarias:** Se ha transformado la columna Salari_mensual a texto para eliminar de forma segura el símbolo de euro (€), los puntos de miles y comas.

**Filtrado de registros incompletos (dropna):** Se ha aplicado un filtro de eliminación de filas basado en la columna de la ratio para descartar automáticamente aquellos registros donde faltaban datos de origen (NA), evitando la propagación de errores en la tabla final.

**Visualización:** Se ha utilizado la función display() mostrando únicamente las primeras 20 filas de las columnas clave (Nom, Cognoms, Salari_mensual, Cost_vida_index y Ratio_Salari_CostVida) para comprobar visualmente con la máxima claridad los datos obtenidos (sin un exceso de datos).


**2. Indicador de dependència familiar (Salari / Nombre de fills)**

**Fórmula principal:** Indicador Dependencia Familiar = Salari Mensual / Nombre Fills

In [118]:
# Convertir las columnas a tipo numérico por si contienen valores de texto
df_enriquit['Nombre_fills'] = pd.to_numeric(
    df_enriquit['Nombre_fills'], errors='coerce'
)
df_enriquit['Salari_mensual'] = pd.to_numeric(
    df_enriquit['Salari_mensual'], errors='coerce'
)

# Creación de la variable derivada sumando 1 al denominador y redondeando a 2 decimales
df_enriquit['Indicador_Dependencia_Familiar'] = (
    df_enriquit['Salari_mensual'] / (df_enriquit['Nombre_fills'] + 1)
).round(2)

# Mostrar una pequeña muestra para comprobar el resultado
display(
    df_enriquit[
        [
            'Nom',
            'Cognoms',
            'Salari_mensual',
            'Nombre_fills',
            'Indicador_Dependencia_Familiar',
        ]
    ].head(20)
)

,Nom,Cognoms,Salari_mensual,Nombre_fills,Indicador_Dependencia_Familiar
0,Joana,Gil Navarro,1469,3,367.25
1,Marc,Muñoz,2718,0,2718.0
2,Noa,Serra,1358,4,271.6
3,Pol,Gil,1478,2,492.67
4,David,Vila,1284,<NA>,<NA>
5,Alexia,Puig,2069,0,2069.0
6,David,Gómez Vila,2349,<NA>,<NA>
7,Carmen,López,2855,4,571.0
8,Nora,Andersen Pedersen,1338,0,1338.0
9,Alba,García,1450,4,290.0


**La seva utilitat analítica.**

El indicador de dependencia familiar permite evaluar la capacidad adquisitiva real y la presión económica individualizada dentro del hogar, poniendo en relación el salario mensual con el número de cargas familiares reales.

Es decir, facilita la detección de perfiles que, a pesar de recibir un salario moderado, sufren un riesgo elevado de estrés económico debido a un volumen elevado de descendencia

**Les decisions preses.**

**Inclusión de la unidad en el denominador (+ 1):** Se ha sumado una unidad al número de hijos (Nombre_fills + 1) como parte de la fórmula de reparto. Esta decisión responde a una doble finalidad.

- Evitar errores técnicos: Previene de forma nativa cualquier posible división por cero en aquellos individuos o perfiles que no cuentan con descendencia.

- Reflejar la unidad real: Contempla a la propia persona trabajadora dentro de la estructura familiar. De este modo, si un empleado tiene un único hijo, el salario se divide entre dos, reflejando que los recursos económicos deben cubrir tanto al proveedor familiar asó como a la persona dependiente.



**3. Índex de mobilitat: Km anuals / Temperatura mitjana**

**Fórmula:** Index Mobilidad = Km Anuales / temperatura Mediana Ciudad

In [119]:
# 1. Forzamos la conversión a numérico de ambas columnas por si son de tipo string
df_enriquit['Km_anuals'] = pd.to_numeric(
    df_enriquit['Km_anuals'], errors='coerce'
)
df_enriquit['Temperatura_mitjana_ciutat'] = pd.to_numeric(
    df_enriquit['Temperatura_mitjana_ciutat'], errors='coerce'
)

# Eliminar filas duplicadas basandome en el DNI
df_enriquit = df_enriquit.drop_duplicates(subset=['DNI'], keep='first')

# 2. Creación de la operación aritmética
df_enriquit['Index_Mobilitat'] = (
    df_enriquit['Km_anuals'] / df_enriquit['Temperatura_mitjana_ciutat']
).round(2)

# 3. Filtrar para eliminar los registros que se hayan quedado sin valor
df_filtrat_mob = df_enriquit.dropna(subset=['Index_Mobilitat'])

# 4. Resultado
display(
    df_filtrat_mob[
        [
            'Nom',
            'Cognoms',
            'Km_anuals',
            'Temperatura_mitjana_ciutat',
            'Index_Mobilitat',
        ]
    ].head(20)
)

,Nom,Cognoms,Km_anuals,Temperatura_mitjana_ciutat,Index_Mobilitat
0,Joana,Gil Navarro,32108,10.1,3179.01
1,Marc,Muñoz,19496,18.7,1042.57
6,David,Gómez Vila,6196,17.8,348.09
7,Carmen,López,20000,15.4,1298.7
10,Joana,Torres Díaz,14178,38.0,373.11
12,Aina,Pérez,10800,22.0,490.91
13,Emilia,Hoffmann,6931,12.4,558.95
14,Xavier,Casals,38248,19.7,1941.52
15,Carmen,Hernández Serra,65950,17.0,3879.41
17,Sigrid,Olsen,80000,8.8,9090.91


**La seva utilitat analítica.**

Este valor permite normalizar el esfuerzo de desplazamiento teniendo en cuenta el factor climático, facilitando la comparación de la movilidad entre trabajadores independientemente de la ciudad en la que residan.

**Les decisions preses.**

**Operación directa y limpieza de valores nulos (dropna):** Dado que no se parte de temperaturas medias iguales a cero, se ha optado por una división directa y se ha aplicado un método .dropna(subset=['Index_Mobilitat']). Esta decisión asegura que el dataset contenga registros completos, eliminando las filas incompletas para así disponer de una muestra fiable para su interpretación.

**Redondeo:** Se ha limitado el resultado final a dos decimales (.round(2)) para facilitar la lectura y comparativa entre los registros.

#Exercici 4: Exportació i síntesi de resultats


**» Utilitzant un bucle, exportar en 4 fitxers (format .csv o .xlsx) les dades corresponents a cada Grup Professional. Per exemple: dades_GrupA.csv, dades_GrupB.csv...**


In [120]:
# Obtenim els grups professionals
grups = df_enriquit['Grup_Professional'].unique()

# Exportem les dades de cada grup en un fitxer CSV
for grup in grups:
    dades_grup = df_enriquit[df_enriquit['Grup_Professional'] == grup]

    nom_fitxer = f"dades_{grup}.csv"

    dades_grup.to_csv(
        nom_fitxer,
        index=False,
        encoding='utf-8-sig'
    )

**» Crear i exportar un cinquè DataFrame (format .csv o .xlsx) que contingui, per a cada Grup Professional:**

**Nombre total de treballadors.**

**Salari mitjà.**

**Edat mediana.**


In [121]:
# Calcular la edad de los trabajadores
df_enriquit['Edat'] = 2026 - pd.to_numeric(
    df_enriquit['Any_de_Naixement'],
    errors='coerce'
)

In [122]:
# Crear el DataFrame resumen para Grupo Professional
df_resum = (
    df_enriquit
    .groupby('Grup_Professional')
    .agg(
        Nombre_treballadors=('Grup_Professional', 'size'),
        Salari_mitja=('Salari_mensual', 'mean'),
        Edat_mediana=('Edat', 'median')
    )
    .reset_index()
)

...y finalmente exportamos

In [123]:
# Exportar el DataFrame resumen
df_resum.to_csv(
    'resum_Grups_Professionals.csv',
    index=False,
    encoding='utf-8-sig'
)

...y mostramos

In [124]:
df_resum['Salari_mitja'] = df_resum['Salari_mitja'].round(2)

display(df_resum)

df_resum.to_csv(
    'resum_Grups_Professionals.csv',
    index=False,
    encoding='utf-8-sig'
)

,Grup_Professional,Nombre_treballadors,Salari_mitja,Edat_mediana
0,Grup A,459,1111.89,44.0
1,Grup B,350,1650.55,45.0
2,Grup C,140,2373.72,48.0
3,Grup D,40,3166.62,39.0


**» Quin nivell d'agrupació és més adequat per a l'exportació de resultats?**


Creo que agrupar los datos por Grupo Profesional es lo más adecuado, ya que permite separar fácilmente a los trabajadores y comparar los diferentes grupos. Además, así podemos tener tanto los datos de cada trabajador como un resumen general de cada grupo.

**» Com separar dades detallades de dades agregades?**

Los datos detallados son los que contienen toda la información de cada trabajador, y los he separado en un archivo para cada Grupo Profesional. En cambio, los datos agregados son un resumen de cada grupo, donde tenemos el número de trabajadores, el salario medio y la edad mediana.

**3. ¿Qué información es relevante para un uso posterior?**

Para un análisis posterior considero especialmente útiles el número de trabajadores, el salario medio y la edad mediana, porque permiten comparar rápidamente los diferentes grupos profesionales. Los datos detallados también son importantes si posteriormente queremos hacer un análisis más concreto o crear visualizaciones.